luaui/Widgets/gui_spectator_hud.lua

```diff
@@ -0,0 +1,2288 @@
+		end
+	end
+	return amountOfTeams
+end
+
+local function getMetricFromID(id)
+	for _,metric in ipairs(metricsEnabled) do
+		if metric.id == id then
+			return metric
+		end
+	end
+	return nil
+end
+
+local function getMetricChosen()
+	for _, currentMetric in ipairs(metricsEnabled) do
+		if metricChosenKey == currentMetric.key then
+			return currentMetric
+		end
+	end
+	return nil
+end
+
+local updateHeaderTooltip -- symbol declaration, function definition later
+local function setMetricChosen(metricKey, ignoreUpdateHeader)
+	local metricChosenKeyOld = metricChosenKey
+	metricChosenKey = metricKey
+	local metricChosen = getMetricChosen()
+	if not metricChosen then
+		metricChosenKey = metricChosenKeyOld
+		return
+	end
+
+	if ignoreUpdateHeader then
+		return
+	end
+
+	headerLabel = metricChosen.title
+	updateHeaderTooltip()
+end
+
+local function buildMetricsEnabled()
+	metricsEnabled = {}
+
+	local index = 1
+	for _,metric in ipairs(metricsAvailable) do
+		if widgetConfig >= metric.configLevel then
+			local metricEnabled = table.copy(metric)
+			metricEnabled.id = index
+			metricsEnabled[index] = metricEnabled
+			if metricChosenKey == metricEnabled.key then
+				metricChosenEnabled = true
+			end
+			local i18nTitleKey = "ui.spectator_hud." .. metricEnabled.key .. "_title"
+			metricEnabled.title = Spring.I18N(i18nTitleKey)
+			local i18nTooltipKey = "ui.spectator_hud." .. metricEnabled.key .. "_tooltip"
+			metricEnabled.tooltip = Spring.I18N(i18nTooltipKey)
+			index = index + 1
+		end
+	end
+
+	if not metricChosenEnabled then
+		local firstAvailableMetric = metricsEnabled[1]
+		setMetricChosen(firstAvailableMetric.key)
+	end
+end
+
+local function getAmountOfMetrics()
+	return #metricsEnabled
+end
+
+local function initMovingAverage(movingAverage)
+	if not settings.useMovingAverage then
+		movingAverage.average = 0
+		return
+	end
+
+	movingAverage.average = 0
+	movingAverage.index = 0
+	movingAverage.data = {}
+	for i=1,settings.movingAverageWindowSize do
+		movingAverage.data[i] = 0
+	end
+end
+
+local function updateMovingAverage(movingAverage, newValue)
+	if not settings.useMovingAverage then
+		movingAverage.average = newValue
+		return
+	end
+
+	if movingAverage.index == 0 then
+		for i=1,settings.movingAverageWindowSize do
+			movingAverage.data[i] = newValue
+		end
+		movingAverage.average = newValue
+		movingAverage.index = 1
+	end
+
+	local newIndex = movingAverage.index + 1
+	newIndex = newIndex <= settings.movingAverageWindowSize and newIndex or 1
+	movingAverage.index = newIndex
+
+	local oldValue = movingAverage.data[newIndex]
+	movingAverage.data[newIndex] = newValue
+
+	movingAverage.average = movingAverage.average + (newValue - oldValue) / settings.movingAverageWindowSize
+
+	if (movingAverage.average * settings.movingAverageWindowSize) < 0.5 then
+		movingAverage.average = 0
+	end
+end
+
+local function getOneStat(statKey, teamID)
+	local result = 0
+
+	if statKey == "metalIncome" then
```

Comment:
Not critical, but for safety from typos and to allow autocomplete, it would be good if these keys were stored in variables in a table. Like
```
keys = {
metalIncome = "metalIncome",
}
```
etc, which can then be accessed with `keys.metalIncome`

Reply:
Done.